recaf-launcher/src/main/java/dev/xdark/recaf/Launcher.java

```diff
@@ -0,0 +1,251 @@
+        try (InputStream in = url.openStream()) {
+          bytes = IOUtils.toByteArray(in);
+        }
+        log("Downloaded updated jar, starting");
+        Files
+            .write(jarPath, bytes, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+        launch(jarPath, Collections.emptyList());
+      } catch (IOException ex) {
+        log("Unable to update jar file, is path writeable? ", ex);
+        if (exists) {
+          attemptLaunch(jarPath, UPDATE_FAIL, 1);
+        }
+      }
+    } else {
+      log("Current jar seems to be valid, using it");
+      attemptLaunch(jarPath, Collections.emptyList());
+    }
+  }
+
+  private static void attemptLaunch(Path jar, List<String> extraOptions, int exit) {
+    try {
+      launch(jar, extraOptions);
+      System.exit(exit);
+    } catch (IOException ex) {
+      log("Unable to launch Recaf: ", ex);
+      System.exit(1);
+    }
+  }
+
+  private static void attemptLaunch(Path jar, List<String> extraOptions) {
+    attemptLaunch(jar, extraOptions, 0);
+  }
+
+  private static void launch(Path jar, List<String> extraOptions) throws IOException {
+    List<String> command = new LinkedList<>(
+        Arrays.asList("java", "-jar", jar.normalize().toString()));
+    command.addAll(extraOptions);
+    new ProcessBuilder()
+        .directory(jar.getParent().toFile())
+        .command(command)
+        .start();
+  }
+
+  private static ReleaseInfo fetchReleaseInfo() throws IOException {
+    URL url = new URL(API_URL);
+    String content;
+    try (InputStream in = url.openStream()) {
+      content = IOUtils.toString(in, StandardCharsets.UTF_8);
+    }
+    ReleaseInfo info = new GsonBuilder()
+        .registerTypeAdapter(Instant.class, new InstantTypeAdapter())
+        .create()
+        .fromJson(content, ReleaseInfo.class);
+    return info;
+  }
+
+  private static void log(String message, Exception ex) {
+    PS.println(message);
+    if (ex != null) {
+      ex.printStackTrace(PS);
+    }
+  }
+
+  private static void log(String message) {
+    log(message, (Exception) null);
+  }
+
+  private static void log(String message, Object... args) {
+    log(String.format(message, args));
+  }
+
+  private static boolean isOutdated(String current, String latest) {
+    String[] arrSelf = current.split("\\.");
+    String[] arrLatest = latest.split("\\.");
+    if (arrSelf.length == arrLatest.length) {
+      for (int i = 0; i < arrSelf.length; i++) {
+        int iSelf = Integer.parseInt(arrSelf[i]);
+        int iLatest = Integer.parseInt(arrLatest[i]);
+        if (iSelf != iLatest) {
+          return iSelf < iLatest;
+        }
+      }
+    }
+    return false;
+  }
+
+  private static Asset findReleaseAsset(List<Asset> assets) {
+    return assets.stream().filter(asset -> asset.getName().matches("recaf-*.*.*.jar")).findFirst()
+        .orElse(null);
+  }
+
+  private static final int VERSION_OFFSET = 44;
+
+  private static int vmVersion() {
```

Comment:
Looks like this is unused?

Reply:
It will be used later for classpath injection

Reply:
I think if we're launching the jar remotely then there is no reason to inject, unless you mean into the current launcher process.